File: AndroidClient/app/src/main/java/com/mancersoft/litevpn/ConnectionManager.java
package com.mancersoft.litevpn;

import android.net.VpnService;
import android.util.Log;

import com.mancersoft.litevpn.transport.IVpnTransport;
import com.mancersoft.litevpn.transport.Packet;
import com.mancersoft.litevpn.transport.TelegramTransport;
import com.mancersoft.litevpn.transport.TransportType;
import com.mancersoft.litevpn.transport.UdpTransport;
import com.mancersoft.litevpn.transport.WebSocketTransport;

import java.io.Closeable;
import java.net.DatagramSocket;
import java.net.Socket;
import java.util.concurrent.CompletableFuture;

import static com.mancersoft.litevpn.VpnManager.MAX_PACKET_SIZE;
import static com.mancersoft.litevpn.VpnManager.PACKET_TYPE_BYTE_OFFSET;
import static com.mancersoft.litevpn.VpnManager.TAG;
import static java.nio.charset.StandardCharsets.US_ASCII;

class ConnectionManager {

    private static final byte CONNECT_PACKET = 0;
    private static final byte DISCONNECT_PACKET = 1;
    private static final int SHARED_SECRET_OFFSET = 1;

    private static final int NOT_RELIABLE_SEND_PARAMS_COUNT = 3;
    private static final int NOT_RELIABLE_HANDSHAKE_ATTEMPTS_COUNT = 50;

    private final String mSharedSecret;
    private final VpnService mService;

    private int mHandshakeAttempt = 0;

    ConnectionManager(String sharedSecret, VpnService vpnService) {
        mSharedSecret = sharedSecret;
        mService = vpnService;
    }

    IVpnTransport createTransport(TransportType transportType, String... transportParams) {
        switch (transportType) {
            case WEBSOCKET:
                return new WebSocketTransport(transportParams[0]);
            case TELEGRAM:
                return new TelegramTransport();
            default:
                return new UdpTransport(transportParams[0], Integer.parseInt(transportParams[1]));
        }
    }

    CompletableFuture<ConnectionParams> sendConnectQuery(IVpnTransport transport) {
        CompletableFuture<ConnectionParams> result = new CompletableFuture<>();
        Packet packet = new Packet();

        byte[] secretBytes = mSharedSecret.getBytes(US_ASCII);
        if (secretBytes.length + 1 > MAX_PACKET_SIZE) {
            completeConnection(transport, result, null);
            return result;
        }

        byte[] connectionData = new byte[secretBytes.length + 1];
        connectionData[PACKET_TYPE_BYTE_OFFSET] = CONNECT_PACKET;
        System.arraycopy(secretBytes, 0, connectionData, SHARED_SECRET_OFFSET, secretBytes.length);
        packet.setData(connectionData);
        packet.setLength(connectionData.length);
        mHandshakeAttempt = 0;
        transport.setOnMessageListener((receivedPacket) -> {
            mHandshakeAttempt++;
            byte[] data = receivedPacket.getData();
            if (data[0] == 0) {
                ConnectionParams params = Utils.getConnectionParams(
                        new String(data, 1, receivedPacket.getLength() - 1, US_ASCII).trim());
                completeConnection(transport, result, params);
            } else if (transport.isReliable()) {
                completeConnection(transport, result, null);
            }

            if (mHandshakeAttempt >= NOT_RELIABLE_HANDSHAKE_ATTEMPTS_COUNT) {
                completeConnection(transport, result, null);
            }
        });
        sendQuery(transport, packet);

        return result;
    }

    void sendDisconnectQuery(IVpnTransport transport) {
        Packet packet = new Packet();
        byte[] data = new byte[PACKET_TYPE_BYTE_OFFSET + 1];
        data[PACKET_TYPE_BYTE_OFFSET] = DISCONNECT_PACKET;
        packet.setData(data);
        packet.setLength(data.length);
        sendQuery(transport, packet);
    }

    private static void sendQuery(IVpnTransport transport, Packet packet) {
        int sendCount = transport.isReliable() ? 1 : NOT_RELIABLE_SEND_PARAMS_COUNT;
        for (int i = 0; i < sendCount; ++i) {
            transport.sendAsync(packet);
        }
    }

    private void completeConnection(IVpnTransport transport,
                                    CompletableFuture<ConnectionParams> result, ConnectionParams value) {
        mHandshakeAttempt = 0;
        transport.setOnMessageListener(null);
        result.complete(value);
    }

    boolean protectTransport(IVpnTransport transport) {
        try {
            Closeable socket = transport.createSocket();
            boolean isProtected;
            if (socket instanceof Socket) {
                isProtected = mService.protect((Socket) socket);
            } else {
                isProtected = mService.protect((DatagramSocket) socket);
            }

            if (!isProtected) {
                throw new IllegalStateException("Cannot protect the tunnel");
            }

            return true;
        } catch (Exception e) {
            Log.e(TAG, "ConnectionManager protectTransport error", e);
            return false;
        }
    }
}

File: AndroidClient/app/src/main/kotlin/com/mancersoft/litevpn/ConnectionManager.kt
package com.mancersoft.litevpn

import android.net.VpnService
import android.util.Log
import com.mancersoft.litevpn.transport.*
import java.net.DatagramSocket
import java.net.Socket
import java.nio.charset.StandardCharsets
import java.util.*
import java.util.concurrent.CompletableFuture

private const val CONNECT_PACKET: Byte = 0
private const val DISCONNECT_PACKET: Byte = 1
private const val SHARED_SECRET_OFFSET = 1
private const val NOT_RELIABLE_SEND_PARAMS_COUNT = 3
private const val NOT_RELIABLE_HANDSHAKE_ATTEMPTS_COUNT = 50

class ConnectionManager(private val mSharedSecret: String, private val mService: VpnService) {

    private var mHandshakeAttempt = 0
    private val mGroupId = UUID.randomUUID().toString()

    private fun sendQuery(transport: IVpnTransport, packet: Packet) {
        val sendCount = if (transport.isReliable) 1 else NOT_RELIABLE_SEND_PARAMS_COUNT
        for (i in 0 until sendCount) {
            transport.sendAsync(packet)
        }
    }

    fun createTransport(transportType: TransportType, vararg transportParams: String): IVpnTransport {
        return when (transportType) {
            TransportType.WEBSOCKET -> WebSocketTransport(transportParams[0])
            else -> UdpTransport(transportParams[0], transportParams[1].toInt())
        }
    }

    fun sendConnectQuery(transport: IVpnTransport): CompletableFuture<ConnectionParams?> {
        val result = CompletableFuture<ConnectionParams?>()
        val packet = Packet()
        val connectDataBytes = ("$mSharedSecret;$mGroupId").toByteArray(StandardCharsets.UTF_8)
        if (connectDataBytes.size + 1 > MAX_PACKET_SIZE) {
            completeConnection(transport, result, null)
            return result
        }
        val connectionData = ByteArray(connectDataBytes.size + 1)
        connectionData[PACKET_TYPE_BYTE_OFFSET] = CONNECT_PACKET
        System.arraycopy(connectDataBytes, 0, connectionData, SHARED_SECRET_OFFSET, connectDataBytes.size)
        packet.data = connectionData
        packet.length = connectionData.size
        mHandshakeAttempt = 0
        transport.setOnMessageListener { receivedPacket: Packet ->
            mHandshakeAttempt++
            val data = receivedPacket.data
            if (data[0] == 0.toByte()) {
                val params = Utils.getConnectionParams(
                        String(data, 1, receivedPacket.length - 1, StandardCharsets.US_ASCII).trim { it <= ' ' })
                completeConnection(transport, result, params)
            } else if (transport.isReliable) {
                completeConnection(transport, result, null)
            }
            if (mHandshakeAttempt >= NOT_RELIABLE_HANDSHAKE_ATTEMPTS_COUNT) {
                completeConnection(transport, result, null)
            }
        }
        sendQuery(transport, packet)
        return result
    }

    fun sendDisconnectQuery(transport: IVpnTransport) {
        val packet = Packet()
        val data = ByteArray(PACKET_TYPE_BYTE_OFFSET + 1)
        data[PACKET_TYPE_BYTE_OFFSET] = DISCONNECT_PACKET
        packet.data = data
        packet.length = data.size
        sendQuery(transport, packet)
    }

    private fun completeConnection(transport: IVpnTransport,
                                   result: CompletableFuture<ConnectionParams?>, value: ConnectionParams?) {
        mHandshakeAttempt = 0
        transport.setOnMessageListener(null)
        result.complete(value)
    }

    fun protectTransport(transport: IVpnTransport): Boolean {
        return try {
            val socket = transport.createSocket()
            val isProtected: Boolean
            isProtected = if (socket is Socket) {
                mService.protect(socket)
            } else {
                mService.protect(socket as DatagramSocket)
            }
            check(isProtected) { "Cannot protect the tunnel" }
            true
        } catch (e: Exception) {
            Log.e(TAG, "ConnectionManager protectTransport error", e)
            false
        }
    }
}